### ISSUE TITLE
Performance regression Wacky Wheels introduction

### ISSUE THREAD

**Issue by jschwartzenberg** [stsp/dosemu2#463]
The intro was always relatively slow, but now I noticed it became even slower.
I bisected it to this commit:
f24a0266e91497ab0c118e76ba171c3d65ec9049 is the first bad commit
commit f24a0266e91497ab0c118e76ba171c3d65ec9049
Author: Stas Sergeev <stsp@users.sourceforge.net>
Date:   Thu Jul 6 19:30:55 2017 +0300

    various: stop using dpmi_base [#385]

:040000 040000 4f4d6a88da893f296273ea154861aa71837706df 5537c62c3b2ed7f1f4c44d62851a5f3b2c01622e Msrc

The whoosh sound now plays long after the music has already finished in the Apogee logo screen.

**Comment by jschwartzenberg** [comment]
It indeed works great again now! Thanks!!

**Comment by stsp** [comment]
Not for me though.
I've tested now myself.
The real regression happens here:
1bdf3749

**Comment by jschwartzenberg** [comment]
I guess there were multiple regressions and I didn't realize it was even faster before!

**Comment by stsp** [comment]
Do you confirm my finding?

**Comment by jschwartzenberg** [comment]
The game crashes here with 1bdf374^, but the part that works appears to be even more smooth indeed.

**Comment by stsp** [comment]
Unfortunately I don't know how this
can be fixed w/o breaking goblins3
again. Maybe @bartoldeman can
help?

**Comment by stsp** [comment]
Perhaps the optimization can be returned
if in addition to `(vga.seq.map_mask | 0xf8) & (vga.seq.map_mask - 1) == 0`
we also check that read and write modes
are 0 (GMR), rotate reg is 0, Enable Set/Reset
reg is 0, bit mask reg=0xff. This combination
seems to allow to safely handle the planar mode
with just mmap()s as the external logic is all
disabled and only 1 plane enabled.
But I wonder if it worth the trouble or to just
leave wacky "broken"?
Also it seems that "vm86sim" performs better
than jit with wacky, which probably means that
cpatch doesn't happen where needed.

**Comment by stsp** [comment]
Ok, so cpatch doesn't happen because the
faults come from DPMI. What matters is that
instremu knows too few commands and just
always exits back to DOS after simulating the
faulted one. This leads to the too frequent
faults in instremu mode.
Perhaps I won't fix this.

**Comment by stsp** [comment]
Its so nice to play wacky wheels at
that speed: you always win! :)
And on dosbox it is blazingly fast,
much more difficult to play. :)
Damn.
It would be a real shame to leave
this unfixed.

**Comment by jschwartzenberg** [comment]
What if you play in VGA passthough mode? I would definitely expect DOSEMU to outperform DOSBox then.

**Comment by stsp** [comment]
> I would definitely expect

You would only expect because you can't even
easily test that. This is not interesting, no one
uses pass-through.

**Comment by jschwartzenberg** [comment]
But I guess it would help towards finding a solution. I recently switched from Ubuntu 14.04 to OpenSUSE 42.3 on this system. It appears it's using KMS for the first time. I never had much issues with pass-through before, on my other systems it should still work.

**Comment by stsp** [comment]
OK, lets see if it is fast enough with pass-through.

**Comment by stsp** [comment]
So Wacky uses mode-X with read mode 0
and write mode 1. Write mode 1 uses latch
but Wacky uses movsb for copying sprites,
so it writes the same value it have read.
That value is equal to the value of the
latch, so on dosemu1 it works fine with
video mem unprotected.
In dosemu2 I don't see how can we
unprotect the video mem for that
setup of mode-X. Likely some other
optimization have to be added, but
there is nothing simple that come in
mind.

**Comment by stsp** [comment]
In fact, the simple fix does exist.
It appears that on a non-debug kernels
the performance is at least adequate.
So all we need is to find the kernel
debug option that slows things down
and ask fedora maintainers to please
disable it even for debug builds.
Since everyone here uses ubuntu, I
don't suppose someone wants to do
the searching? :)

**Comment by jschwartzenberg** [comment]
Btw. is this regression related to kvm or does it happen with vm86 or emulated as well?

**Comment by stsp** [comment]
Absolutely everything.
This is because it happens in DPMI,
so the emu of vm86 does not play any
role. Maybe the real fix would be to
extend instremu into a yet another
CPU emulator (how many of those
do we have there to the date?), but
I'll stick to looking up the nasty kernel
debug option that makes it very visible.

**Comment by stsp** [comment]
Bart, can we fix this by replacing instremu
with simx86? The problem here is that
instremu does not know most instructions,
so in mode-X it just simulates the memory
access, and then, seeing the unknown
instruction ahead, gives up and goes for
another fault.
As fullsim now works properly, I wonder
if it can be used instead: it knows all
instructions, so it can drive mode-X in a
mostly faultless manner.

**Comment by bartoldeman** [comment]
This is an old idea (almost ~20 years old!) and I remember Alberto saying that it should be resolved at a high level, i.e. temporarily enabling cpuemu and calling e_vm86/e_dpmi at the top level.

I can work on that later.

**Comment by stsp** [comment]
Would be cool!
Wacky is a remarkable game, so
having it slower than on both dosbox
and dosemu1 is a big shame.
And I won't miss instremu too much.
The instr_dec library that I derived
from it, can also be replaced with udis86.
So I just think its a good time to remove
instremu.

> a high level, i.e. temporarily enabling cpuemu and calling e_vm86/e_dpmi at the top level.

I more prefer the current approach, where
you do things locally, and not spread the
vgaemu problems around too widely.
But I can imagine that resolving it at a
high level may mean better performance
and fewer faults.
So neither way suits the best.

Hmm. Just a bit of brain-storming.
Can we have a mix of native+jit execution?
I mean, you can mark some code pages
NX, and on those - switch to the compiled
code, which will have mode-X accesses
cpatched. This will likely be blazingly fast,
i.e. we could over-speed dosemu1 here,
which didn't care to protect mode-X planes
at all.

**Comment by stsp** [comment]
doom2 suffers the same problem.

**Comment by stsp** [comment]
Hmm, both Doom and Wacky seem
to set write mode to 1 only temporary.
They mostly use write mode 0.
Which is good as perhaps that case
can be optimized.

**Comment by stsp** [comment]
Found a bug.
Unfortunately the proper
optimization doesn't look
easy. Some progs (goblins 3)
toggle write mode between
0 and 1 too frequently. We
can't commence instremu
too quickly because when
dirty pages are still unprotected,
instremu is not going to work...

In theory someone can switch
between planar/linear modes
quickly too, so perhaps we
need to flush all dirty pages
before enabling instremu?

**Comment by stsp** [comment]
I now protect all video
pages before starting
instremu. But I still have
not found a way to optimize
mode-x because it has
map mask register set
to 0xf, which means that
all planes are written to.
Maybe it is possible, in the
trivial case, to back up the
original plane and then see
what was changed (like
in vgacpy branch), and by
that reconstruct other planes,
but I don't think this is going
to work even when
read_mode==write_mode==0.

**Comment by stsp** [comment]
Hi Bart, if you run `doom2.exe -devparm` then
you'll see the FPS meter in the bottom-left corner.
It consists of a few very small flashing dots.
The fewer dots are lighting, the better.
Seems dosbox still has a faster mode-X.

**Comment by bartoldeman** [comment]
I see about 2-3 dots on average including the bottom left. DOSEMU in non-debug mode seems a bit faster though, but haven't tested that yet vs dosbox. I can do some more testing later.

**Comment by stsp** [comment]
Under dosbox there seem to be only
the left-most dot on (it probably never
is off), and no dots are flashing. So
that seems to indicate the fastest
performance, even in full-screen mode
(Alt-Enter for dosbox).
But there are multiple problems with
doom2 in dosbox, its quite unusable
here. No doubts dosbox can be
configured, but out-of-the-box config
seems to never work well for me.
But mode-X is very fast there.

**Comment by stsp** [comment]
Should finally be fixed.

**Comment by jschwartzenberg** [comment]
Yep, it works again as I'm used to on dosemu!

Note that the fading in and out is still quite slow compared to bare metal.

**Comment by stsp** [comment]
> Note that the fading in and out is still quite slow compared to bare metal.

Try `$_cpu_vm_dpmi = "native"`,
it seems to be now the fastest...
How is that, given that I moved
it under the searpc-based server
process??

**Comment by jschwartzenberg** [comment]
Oh wow, that seems to run absolutely great! At least for Wacky Wheels!!

**Comment by jschwartzenberg** [comment]
It seems to break Jungle Book though. And unpatched Jazz Jackrabbit is very special...
